### PR TITLE
Add QgsGeometry::normalize()

### PR DESCRIPTION
**Pull request by nyalldawson**
A port of the equivalent method from GEOS, but with added support
for curved geometries and M values

Reorganizes the geometry into a normalized form (or "canonical" form).

Polygon rings will be rearranged so that their starting vertex is
the lower left and ring orientation follows the right hand rule, collections
are ordered by geometry type, and other normalization techniques are applied.
The resultant geometry will be geometrically equivalent to the original geometry.

### CODE REVIEW

File: src/core/geometry/qgsabstractgeometry.h [hunk 1]
```diff
@@ -235,6 +235,17 @@ class CORE_EXPORT QgsAbstractGeometry
      */
     virtual QgsAbstractGeometry *boundary() const = 0 SIP_FACTORY;
 
+    /**
+     * Reorganizes the geometry into a normalized form (or "canonical" form).
+     *
+     * Polygon rings will be rearranged so that their starting vertex is the lower left and ring orientation follows the
+     * right hand rule, collections are ordered by geometry type, and other normalization techniques are applied. The
+     * resultant geometry will be geometrically equivalent to the original geometry.
```
Comment (pblottiere) [inline src/core/geometry/qgsabstractgeometry.h:243]:
```suggestion
     * Resultant geometry will be geometrically equivalent to the original geometry.
```

File: src/core/geometry/qgsgeometry.h [hunk 2]
```diff
@@ -2140,6 +2140,17 @@ class CORE_EXPORT QgsGeometry
      */
     void validateGeometry( QVector<QgsGeometry::Error> &errors SIP_OUT, ValidationMethod method = ValidatorQgisInternal, QgsGeometry::ValidityFlags flags = QgsGeometry::ValidityFlags() ) const;
 
+    /**
+     * Reorganizes the geometry into a normalized form (or "canonical" form).
+     *
+     * Polygon rings will be rearranged so that their starting vertex is the lower left and ring orientation follows the
+     * right hand rule, collections are ordered by geometry type, and other normalization techniques are applied. The
+     * resultant geometry will be geometrically equivalent to the original geometry.
```
Comment (pblottiere) [inline src/core/geometry/qgsgeometry.h:2148]:
```suggestion
     * Resultant geometry will be geometrically equivalent to the original geometry.
```

File: src/core/geometry/qgsabstractgeometry.h (hunk 1 above)
Comment (nyalldawson) [reply src/core/geometry/qgsabstractgeometry.h:243]: It follows along from the "The" on the previous line 😉

Comment (pblottiere) [reply src/core/geometry/qgsabstractgeometry.h:243]: Oops, I missed it. Sorry ^^.

**Pull request merged**